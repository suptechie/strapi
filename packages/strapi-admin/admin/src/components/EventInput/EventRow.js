import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from '@buffetjs/core';

const EventRow = ({
  name,
  events,
  inputValue,
  handleChange,
  handleChangeAll,
}) => {
  const areAllCheckboxesSelected = inputValue.length === 3;
  const hasSomeCheckboxSelected = inputValue.length > 0;

  const onChangeAll = ({ target: { name } }) => {
    const valueToSet = !areAllCheckboxesSelected;

    handleChangeAll({
      target: { name, value: valueToSet },
    });
  };

  return (
    <tr>
      <td>
        <Checkbox
          name={name}
          onChange={onChangeAll}
          message={name}
          someChecked={hasSomeCheckboxSelected && !areAllCheckboxesSelected}
          value={areAllCheckboxesSelected}
        />
      </td>

      {events.map(event => {
        return (
          <td key={event}>
            <Checkbox
              name={event}
              value={inputValue.includes(event)}
              onChange={handleChange}
            />
          </td>
        );
      })}
    </tr>
  );
};

EventRow.defaultProps = {
  events: [],
  inputValue: [],
  handleChange: () => {},
  handleChangeAll: () => {},
};

EventRow.propTypes = {
  events: PropTypes.array,
  inputValue: PropTypes.array,
  handleChange: PropTypes.func,
  handleChangeAll: PropTypes.func,
  name: PropTypes.string.isRequired,
};

export default EventRow;
