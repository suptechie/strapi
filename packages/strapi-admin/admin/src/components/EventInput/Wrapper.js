/**
 *
 * Wrapper
 *
 */

import styled from 'styled-components';

const Wrapper = styled.div`
  padding-top: 3px;
  padding-bottom: 8px;
  table {
    width: 100%;
    border-radius: 3px;
    overflow: hidden;
  }
  tr {
    &:before {
      content: '-';
      display: inline-block;
      line-height: 1.1em;
      color: transparent;
      background-color: #f6f6f6;
      position: absolute;
      left: 20px;
      width: calc(100% - 40px);
      height: 1px;
      margin-top: -1px;
    }
    &:first-of-type:before {
      background-color: #fafafb;
      z-index: 1;
      height: 2px;
      width: 100%;
      left: 0;
    }
  }
  thead {
    background-color: #fafafb;
    td {
      height: 41px;
      text-align: center;
      font-size: 13px;
      font-weight: 500;
    }
  }
  tbody {
    border-bottom-left-radius: 3px;
    border-bottom-right-radius: 3px;
    box-shadow: inset 0px 0px 0px 1px #f6f6f6;

    td {
      height: 54px;
      padding-bottom: 3px;
      &:first-of-type {
        padding-left: 30px;
        text-transform: capitalize;
        width: 200px;
        label {
          width: fit-content;
          margin-top: -1px;
          font-weight: 500;
          cursor: pointer;
        }
      }
      &:not(:first-of-type) > div {
        margin: 0 auto;
        display: block;
        width: fit-content;
      }
    }
  }
`;

export default Wrapper;
