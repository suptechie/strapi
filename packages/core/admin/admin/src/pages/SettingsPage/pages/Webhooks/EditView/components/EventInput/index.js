import React from 'react';
import PropTypes from 'prop-types';
import { FieldLabel } from '@strapi/design-system/Field';
import { Stack } from '@strapi/design-system/Stack';
import { P, TableLabel } from '@strapi/design-system/Text';
import { useFormikContext } from 'formik';
import { useIntl } from 'react-intl';
import styled from 'styled-components';
import EventRow from './EventRow';
import formatValue from './utils/formatValue';

const StyledTable = styled.table`
  td {
    height: ${52 / 16}rem;
    width: 10%;
    vertical-align: middle;
    text-align: center;
  }

  tbody tr:nth-child(odd) {
    background: ${({ theme }) => theme.colors.neutral100};
  }

  tbody tr td:first-child {
    padding-left: ${({ theme }) => theme.spaces[7]};
  }
`;

const displayedData = {
  headers: {
    default: [
      'Settings.webhooks.events.create',
      'Settings.webhooks.events.update',
      'app.utils.delete',
    ],
    draftAndPublish: [
      'Settings.webhooks.events.create',
      'Settings.webhooks.events.update',
      'app.utils.delete',
      'app.utils.publish',
      'app.utils.unpublish',
    ],
  },
  events: {
    default: {
      entry: ['entry.create', 'entry.update', 'entry.delete'],
      media: ['media.create', 'media.update', 'media.delete'],
    },
    draftAndPublish: {
      entry: ['entry.create', 'entry.update', 'entry.delete', 'entry.publish', 'entry.unpublish'],
      media: ['media.create', 'media.update', 'media.delete'],
    },
  },
};

const EventInput = ({ isDraftAndPublish }) => {
  const headersName = isDraftAndPublish
    ? displayedData.headers.draftAndPublish
    : displayedData.headers.default;

  const events = isDraftAndPublish
    ? displayedData.events.draftAndPublish
    : displayedData.events.default;

  const { formatMessage } = useIntl();
  const { values, errors, handleChange: onChange } = useFormikContext();

  const inputName = 'events';
  const inputValue = values.events;

  const disabledEvents = [];

  const formattedValue = formatValue(inputValue);

  const handleChange = ({ target: { name, value } }) => {
    let set = new Set(inputValue);

    if (value) {
      set.add(name);
    } else {
      set.delete(name);
    }
    onChange({ target: { name: inputName, value: Array.from(set) } });
  };

  const handleChangeAll = ({ target: { name, value } }) => {
    let set = new Set(inputValue);

    if (value) {
      events[name].forEach(event => {
        if (!disabledEvents.includes(event)) {
          set.add(event);
        }
      });
    } else {
      events[name].forEach(event => set.delete(event));
    }
    onChange({ target: { name: inputName, value: Array.from(set) } });
  };

  return (
    <Stack size={1}>
      <FieldLabel>
        {formatMessage({
          id: 'Settings.webhooks.form.events',
          defaultMessage: 'Events',
        })}
      </FieldLabel>
      <StyledTable>
        <thead>
          <tr>
            <td />
            {headersName.map(header => {
              if (header === 'app.utils.publish' || header === 'app.utils.unpublish') {
                return (
                  <td
                    key={header}
                    title={formatMessage({
                      id: 'Settings.webhooks.event.publish-tooltip',
                      defaultMessage:
                        'This event only exists for contents with Draft/Publish system enabled',
                    })}
                  >
                    <TableLabel textColor="neutral600">{formatMessage({ id: header })}</TableLabel>
                  </td>
                );
              }

              return (
                <td key={header}>
                  <TableLabel textColor="neutral600">{formatMessage({ id: header })}</TableLabel>
                </td>
              );
            })}
          </tr>
        </thead>
        <tbody>
          {Object.keys(events).map(event => {
            return (
              <EventRow
                disabledEvents={disabledEvents}
                key={event}
                name={event}
                events={events[event]}
                inputValue={formattedValue[event]}
                handleChange={handleChange}
                handleChangeAll={handleChangeAll}
              />
            );
          })}
        </tbody>
      </StyledTable>
      {errors.events && (
        <P small textColor="danger600" data-strapi-field-error>
          {formatMessage({
            id: 'components.Input.error.validation.required',
            defaultMessage: 'This value is required',
          })}
        </P>
      )}
    </Stack>
  );
};

EventInput.propTypes = {
  isDraftAndPublish: PropTypes.bool.isRequired,
};

export default EventInput;
