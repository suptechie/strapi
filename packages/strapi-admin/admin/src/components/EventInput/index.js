import React from 'react';
import PropTypes from 'prop-types';
import { useGlobalContext } from 'strapi-helper-plugin';

import Wrapper from './Wrapper';
import EventRow from './EventRow';

const EventInput = ({ onChange, name: inputName, value: inputValue }) => {
  const { formatMessage } = useGlobalContext();

  const headersName = [
    formatMessage({ id: `Settings.webhooks.events.create` }),
    formatMessage({ id: `Settings.webhooks.events.edit` }),
    formatMessage({ id: `Settings.webhooks.events.delete` }),
  ];

  const events = {
    entry: ['createEntry', 'updateEntry', 'deleteEntry'],
    media: ['createMedia', 'updateMedia', 'deleteMedia'],
  };

  const handleChange = ({ target: { name, value } }) => {
    let newValue = inputValue;

    if (value) {
      if (!inputValue.includes(name)) {
        newValue.push(name);
      }
    } else {
      if (inputValue.includes(name)) {
        const index = inputValue.indexOf(name);
        if (index > -1) {
          newValue.splice(index, 1);
        }
      }
    }

    onChange({ target: { name: inputName, value: newValue } });
  };

  const handleChangeAll = ({ target: { name, value } }) => {
    let newValue = inputValue;

    if (value) {
      events[name].map(event => {
        if (!inputValue.includes(event)) {
          newValue.push(event);
        }
      });
    } else {
      events[name].map(event => {
        const index = inputValue.indexOf(event);
        if (index > -1) {
          newValue.splice(index, 1);
        }
      });
    }

    onChange({ target: { name: inputName, value: newValue } });
  };

  return (
    <Wrapper>
      <table>
        <thead>
          <tr>
            <td></td>
            {headersName.map(header => {
              return <td key={header}>{header}</td>;
            })}
          </tr>
        </thead>
        <tbody>
          {Object.keys(events).map(event => {
            return (
              <EventRow
                key={event}
                name={event}
                events={events[event]}
                inputValue={inputValue}
                handleChange={handleChange}
                handleChangeAll={handleChangeAll}
              />
            );
          })}
        </tbody>
      </table>
    </Wrapper>
  );
};

EventInput.defaultProps = {
  handleClick: () => {},
  onClick: () => {},
};

EventInput.propTypes = {
  handleClick: PropTypes.func,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
  value: PropTypes.array,
};

export default EventInput;
