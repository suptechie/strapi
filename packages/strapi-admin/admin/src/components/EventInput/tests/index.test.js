import React from 'react';
import renderer from 'react-test-renderer';
import { IntlProvider } from 'react-intl';

import EventInput from '../index';

describe('<EventInput />', () => {
  const props = {
    name: 'events',
    value: ['media.create, media.delete'],
    onChange: jest.fn(),
  };

  it('should render properly', () => {
    const tree = renderer.create(
      <IntlProvider locale="en">
        <EventInput {...props} />
      </IntlProvider>
    );

    expect(tree).toMatchSnapshot();
  });
});
