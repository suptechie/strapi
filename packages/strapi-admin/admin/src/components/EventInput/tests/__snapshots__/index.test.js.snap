// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EventInput /> should match the snapshot 1`] = `
<div
  className="sc-jAaTju cbZcAq"
>
  <table>
    <thead>
      <tr>
        <td />
        <td>
          <span>
            Create
          </span>
        </td>
        <td>
          <span>
            Edit
          </span>
        </td>
        <td>
          <span>
            Delete
          </span>
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              id="entry"
              name="entry"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="entry"
            >
              entry
            </label>
          </div>
        </td>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              disabled={false}
              id="entry.create"
              name="entry.create"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="entry.create"
            />
          </div>
        </td>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              disabled={false}
              id="entry.update"
              name="entry.update"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="entry.update"
            />
          </div>
        </td>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              disabled={false}
              id="entry.delete"
              name="entry.delete"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="entry.delete"
            />
          </div>
        </td>
      </tr>
      <tr>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat cllrAw"
              id="media"
              name="media"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="media"
            >
              media
            </label>
          </div>
        </td>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              disabled={false}
              id="media.create"
              name="media.create"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="media.create"
            />
          </div>
        </td>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              disabled={true}
              id="media.update"
              name="media.update"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="media.update"
            />
          </div>
        </td>
        <td>
          <div
            className="sc-bxivhb hRdlIh"
          >
            <input
              autoComplete="off"
              autoFocus={false}
              checked={false}
              className="sc-htpNat BNnRX"
              disabled={false}
              id="media.delete"
              name="media.delete"
              onChange={[Function]}
              placeholder={null}
              tabIndex="0"
              type="checkbox"
            />
            <label
              className="sc-chPdSV iXihgI"
              htmlFor="media.delete"
            />
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
